onnxruntime/core/providers/cuda/tensor/scatter_nd.h

```diff
@@ -3,18 +3,63 @@
 
 #pragma once
 
+#include <string>
 #include "core/providers/shared_library/provider_api.h"
 #include "core/providers/cuda/cuda_kernel.h"
+#include "core/providers/cuda/tensor/scatter_nd_kind.h"
 #include "core/providers/cpu/tensor/scatter_nd.h"
 
 namespace onnxruntime {
 namespace cuda {
 
-class ScatterND final : public CudaKernel {
+/**
+ * This implementation assumes there is common indices and
+ * reduction is not needed. The code does not check that condition.
+ * However in that case, the same output element could be accessed
+ * from different threads at the same time and the final value
+ * is unlikely to be correct.
+ */
+class ScatterNDDisjointAndNoReduction final : public CudaKernel {
  public:
-  explicit ScatterND(const OpKernelInfo& info) : CudaKernel(info) {}
+  explicit ScatterNDDisjointAndNoReduction(const OpKernelInfo& info) : CudaKernel(info) {}
   Status ComputeInternal(OpKernelContext* context) const override;
 };
 
+/**
+ * This is an implementation derived from the first one.
+ * It does atomic operation to handle conflicts.
+ * The result is unlikely to be correct if the reduction is none
+ * as there is no guarantee that the final value will be the one
+ * corresponding to the highest visited index.
+ * TODO: change the implementation of avoid conflicts.
+ */
+class ScatterNDWithAtomicReduction final : public CudaKernel {
+ public:
+  explicit ScatterNDWithAtomicReduction(const OpKernelInfo& info) : CudaKernel(info) {
+    std::string reduction = info.GetAttrOrDefault<std::string>("reduction", "none");
+
+    if (info.GetAttr<std::string>("reduction", &reduction).IsOK()) {
+      if (reduction == "add") {
+        reduction_ = ScatterNDReduction::Add;
+      } else if (reduction == "mul") {
+        reduction_ = ScatterNDReduction::Mul;
+      } else if (reduction == "min") {
+        reduction_ = ScatterNDReduction::Min;
+      } else if (reduction == "max") {
+        reduction_ = ScatterNDReduction::Max;
+      } else if (reduction == "none") {
```

Comment:
The spec says, indecies should not have duplicated. Are we checking for that? If that is true, the the same output can not be accessed by more than one thread. I think this warning is not needed.

Reply:
I says also: In cases where reduction is set to “none”, indices should not have duplicate entries: that is, if idx1 != idx2, then indices[idx1] != indices[idx2]. I think we forgot to remove the first sentance after adding reduction as an argument. Torch script exporter is using ScatterND to express a torch function which allows duplicates. So I assumed, duplicates are allowed if reduction is not none. We need that.

Reply:
Okay, but they should be unique of reduction is not `none`?

Reply:
I did not change the existing implementation if the reduction is none: https://github.com/microsoft/onnxruntime/pull/19540/files#diff-6b2644c8e90870d9c6069fde07571378dd9f35d3495d37db2c8191716765b1f1R170 and it is not using atomic.